dali/operators/segmentation/random_object_bbox.cc

```diff
@@ -0,0 +1,442 @@
+    if (!classes_.IsDefined()) {
+      classes.resize(ncls);
+      int cls = 0;
+      for (int i = 0; i < ncls; i++, cls++) {
+        if (cls == background)
+          cls++;
+        classes[i] = cls;
+      }
+    }
+  }
+}
+
+template <int ndim>
+int RandomObjectBBox::PickBox(span<Box<ndim, int>> boxes, int sample_idx) {
+  auto beg = boxes.begin();
+  auto end = boxes.end();
+  if (threshold_.IsDefined()) {
+    vec<ndim, int> threshold;
+    const int *thresh = threshold_[sample_idx].data;
+    assert(threshold_.get().shape[sample_idx] == TensorShape<1>{ ndim });
+    for (int i = 0; i < ndim; i++)
+      threshold[i] = thresh[i];
+    end = std::remove_if(beg, end, [threshold](auto box) {
+      return any_coord(box.extent() < threshold);
+    });
+  }
+  int n = end - beg;
+  if (n <= 0)
+    return -1;
+
+  if (k_largest_ > 0) {
+    SmallVector<std::pair<int64_t, int>, 32> vol_idx;
+    vol_idx.resize(n);
+    for (int i = 0; i < n; i++) {
+      vol_idx[i] = { -volume(boxes[i]), i };
+    }
+    std::sort(vol_idx.begin(), vol_idx.end());
+    std::uniform_int_distribution<int> dist(0, std::min(n, k_largest_)-1);
+    return vol_idx[dist(rngs_[sample_idx])].second;
+  } else {
+    std::uniform_int_distribution<int> dist(0, n-1);
+    return dist(rngs_[sample_idx]);
+  }
+}
+
+bool RandomObjectBBox::PickBlob(SampleContext &ctx, int nblobs) {
+  if (!nblobs)
+    return false;
+
+  int ndim = ctx.blobs.dim();
+  ctx.box_data.clear();
+  ctx.box_data.resize(2*ndim*nblobs);
+
+  VALUE_SWITCH(ndim, static_ndim, (1, 2, 3, 4, 5, 6),
+    (
+      auto *box_data = reinterpret_cast<Box<static_ndim, int>*>(ctx.box_data.data());
+      auto boxes = make_span(box_data, nblobs);
+      GetLabelBoundingBoxes(boxes, ctx.blobs.to_static<static_ndim>(), -1);
+      int box_idx = PickBox(boxes, ctx.sample_idx);
+      if (box_idx >= 0) {
+        ctx.SelectBox(box_idx);
+        return true;
+      }
+    ), (  // NOLINT
+      DALI_FAIL(make_string("Unsupported number of dimensions: ", ndim, "; must be 1..6"));
+    )  // NOLINT
+  );  // NOLINT
+  return false;
+}
+
+template <typename T>
+bool RandomObjectBBox::PickForegroundBox(
+      SampleContext &context, const InTensorCPU<T> &input) {
+  GetClassesAndWeightsArgs(context.classes, context.weights, context.background,
+                           context.sample_idx);
+  context.class_label = context.background;
+  if (ignore_class_) {
+    int nblobs = LabelConnectedRegions(context.blobs, input, -1, context.background);
+    return PickBlob(context, nblobs);
+  } else {
+    FindLabels(context.labels, input);
+
+    context.labels.erase(context.background);
+
+    if (!classes_.IsDefined() && !weights_.IsDefined()) {
+      context.classes.clear();
+      context.weights.clear();
+      for (auto cls : context.labels) {
+        context.classes.push_back(cls);
+        context.weights.push_back(1);
+      }
+      std::sort(context.classes.begin(), context.classes.end());
```

Comment:
It's cheaper to copy it from unordered set and sort, because unordered set has better memory management than `set` and we need a copy to a flat array anyway.

Reply:
I would be for writing such comments directly in the code instead of the PR review.